src/lsm/segmented_array.zig

```diff
@@ -33,10 +34,17 @@ pub fn SortedSegmentedArray(
     comptime Key: type,
     comptime key_from_value: fn (*const T) callconv(.Inline) Key,
     comptime compare_keys: fn (Key, Key) callconv(.Inline) math.Order,
+    comptime options: Options,
 ) type {
-    return SegmentedArrayType(T, NodePool, element_count_max, Key, key_from_value, compare_keys);
+    return SegmentedArrayType(T, NodePool, element_count_max, Key, key_from_value, compare_keys, options);
 }
 
+pub const Options = struct {
```

Comment:
What do you think of renaming `Options` to `Config`? Typically we use `Options` for runtime options passed to an `init()` function, whereas "config" is comptime.

Reply:
Sure.

Reply:
@sentientwaffle I think let's actually keep `Options`, whether runtime or comptime (i.e. not coupled to one or the other), also since `config.zig` will be renamed to `constants.zig`.

Reply:
I've actually been thinking that `config.zig` could stay `config.zig`, and it would look something like:

```
pub const UserConfig = struct {
    /// description of this config option
    message_size_max: usize = 1024 * 1024,
    /// description
    journal_slot_count: usize = 1024,
    // ... and all of the other user-configurable options
};

pub fn Config(comptime user_config: UserConfig) type {
    // verify the user's config is all valid:
    assert(...);
    // ... etc

    return struct {
        // includes the user's configuration:
        pub const message_size_max = user_config.message_size_max;
        pub const journal_slot_count = user_config.journal_slot_count;
        // ...

        // includes constants
        pub const sector_size = 4096;

        // also includes any constants that are derived from the user's config
        pub const journal_size_prepares = journal_slot_count * message_size_max;
    };
}

pub fn config_from_build_options(defaults: UserConfig) type {
    // returns a Config(...) where the "user_config"'s values are provided by "zig build" options (see Builder.addOptions).
}
```

- `config_from_build_options()` would be called from `main.zig`
- Simulator
  - `config_from_build_options()` would also be called from `simulator.zig` to allow running the VOPR with different profiles, to find interesting bugs (e.g. vary the LSM growth level & depth, vary the WAL size, vary the message size or clients count, etc). This strategy of fuzzing configs comes from the Swarm Testing paper iirc.
  - Either `vopr.zig` could generate configs, or it could just generate a "config seed" (so we can still reproduce it!) which it passes to `zig build simulator`, and then `build.zig` could generate a config and pass it via `Builder.addOptions()` into the simulator.
  - When running the simulator repeatedly (e.g. in the vopr-hub) we wouldn't change the config seed every single run, rather every 50 or 100 runs — this way we amortize the recompilation cost.
- Our other fuzzers (e.g. of LSM components) could reuse the same config-generation as the simulator, to similar benefit.
  - I think this is especially valuable (or even essential) for testing the LSM. Since I don't have terabytes of memory (or storage) to exhaustively test a full LSM, we can instead shrink the LSM and explore it more deeply.
- Unit testing
  - Sometimes unit tests can be written to be general over valid values of `config`s, but often it is easier to write a known case for a specific configuration.
  - For example, I have a branch that unit tests the SuperBlock `Quorums.working()`, and it specifically targets `superblock_copies=4`. Instead of having that test just fail if run with `superblock_copies=6`, the `Quorums` could take a `comptime Config: type` parameter, and the unit test will test `Quorums(Config(.{ .superblock_copies = 4 }))`.

Types that currently depend on `config` via `const config = @import("config.zig");` should depend on a `comptime config: type` instead of allow this kind of testing.

---

And I guess I got off track. :sweat_smile:  I'm ok with `Options` here, but we may want to reuse this pattern elsewhere for types that already have an `Options`.

Reply:
Tree/Forest in particular will need the Config/Options split if we want to be able to fuzz different configs.

Reply:
If Joran doesn't like Config, maybe ComptimeOptions/Options?

Reply:
Where we have both in scope we can do `OptionsComptime` and `OptionsRuntime`, with the qualifier last.